salt/states/vagrant.py

```diff
@@ -0,0 +1,368 @@
+
+    .. code-block:: yaml
+
+        node_name:
+          vagrant.running
+
+    .. code-block:: yaml
+
+        node_name:
+          vagrant.running:
+            - cwd: /projects/my_project
+            - vagrant_runas: my_username
+            - machine: machine1
+
+    '''
+    if '*' in name or '?' in name:
+
+        return _vagrant_call(name, 'start', 'restarted',
+                             "Machine has been restarted", "running")
+
+    else:
+
+        ret = {'name': name,
+               'changes': {},
+               'result': True,
+               'comment': '{0} is already running'.format(name)
+               }
+
+        try:
+            info = __salt__['vagrant.vm_state'](name)
+            if info[0]['state'] != 'running':
+                __salt__['vagrant.start'](name)
+                ret['changes'][name] = 'Machine started'
+                ret['comment'] = 'Node {0} started'.format(name)
+        except (SaltInvocationError, CommandExecutionError):
+            #  there was no viable existing machine to start
+            ret, kwargs = _find_init_change(name, ret, **kwargs)
+            kwargs['start'] = True
+            __salt__['vagrant.init'](name, **kwargs)
+            ret['changes'][name] = 'Node defined and started'
+            ret['comment'] = 'Node {0} defined and started'.format(name)
+
+        return ret
+
+
+def _find_init_change(name, ret, **kwargs):
+    '''
+    look for changes from any previous init of machine.
+
+    :return: modified ret and kwargs
+    '''
+    kwargs = salt.utils.args.clean_kwargs(**kwargs)
+    if 'vm' in kwargs:
+        kwargs.update(kwargs.pop('vm'))
+    # the state processing eats 'runas' so we rename
+    kwargs['runas'] = kwargs.pop('vagrant_runas', '')
+    try:
+        vm_ = __salt__['vagrant.get_vm_info'](name)
+    except SaltInvocationError:
+        vm_ = {}
+        for key, value in kwargs.items():
+            ret['changes'][key] = {'old': None, 'new': value}
+    if vm_:  # test for changed values
+        for key in vm_:
+            value = vm_[key] or ''  # supply a blank if value is None
+            if key != 'name':  # will be missing in kwargs
+                new = kwargs.get(key, '')
+                if new != value:
+                    if key == 'machine' and new == '':
+                        continue  # we don't know the default machine name
+                    ret['changes'][key] = {'old': value, 'new': new}
+    return ret, kwargs
+
+
+def initialized(name, **kwargs):
+    r'''
+    Defines a new VM with specified arguments, but does not start it.
+
+    :param name: the Salt_id node name you wish your VM to have.
+
+    Each machine must be initialized individually using this function
+    or the "vagrant.running" function, or the vagrant.init execution module call.
+
+    This command will not change the state of a running or paused machine.
+
+    Possible keyword arguments:
+
+    - cwd: The directory (path) containing the Vagrantfile
```

Comment:
If these are the keyword arguments which are needed, is there a reason not to just explicitly declare them in the function interface?

Reply:
I think it may be better in this case to keep them generic, because I pass them around so much.
- I hand them to _find_init_change() to parse for changes so I can report them in the `ret` dictionary.
- vagrant.running uses the same set of kwargs and passes them to _find_init_change()
- I pass them directly on to  module.vagrant.init() where the defaults are defined.
- vagrant.running also passes them to module.vagrant.init()
- module.vagrant.init is the source of truth for the default values.